Eager ConcatMap

akarnokd opened the pull request:
Related discussion in #3017.

davidmoten commented:
Thanks a lot for contributing this one @akarnokd. I've had one look through and no problems jump out at me but I will review further. This will be useful.

davidmoten commented on src/main/java/rx/internal/operators/OperatorEagerConcatMap.java:
```diff
@@ -0,0 +1,299 @@
+            drain();
+        }
+        
+        void drain() {
```
For  readability I'd suggest longer names for variables like d, a, r, e, p.

davidmoten commented on src/main/java/rx/internal/operators/OperatorEagerConcatMap.java:
```diff
@@ -0,0 +1,299 @@
+            }
+            int missed = 1;
+            
+            final AtomicLong p = sharedProducer;
```
p is not being used in the sense of a producer but rather in its sense of the request count. I'd like to see this renamed to `requested` or something similar in keeping with naming in other operators.

davidmoten commented on src/main/java/rx/internal/operators/OperatorEagerConcatMap.java:
```diff
@@ -0,0 +1,299 @@
+                }
+
+                if (!empty) {
+                    long r = p.get();
```
`r` could be `req` say and p gets renamed to `requested`

davidmoten commented on src/main/java/rx/internal/operators/OperatorEagerConcatMap.java:
```diff
@@ -0,0 +1,299 @@
+                }
+                
+                missed = wip.addAndGet(-missed);
+                if (missed == 0) {
```
A unit test that covers both branches of this statement might be nice

akarnokd replied:
Done. See the reentrancy test.

davidmoten commented on src/main/java/rx/internal/operators/OperatorEagerConcatMap.java:
```diff
@@ -0,0 +1,299 @@
+        
+        final EagerOuterSubscriber<?, ?> parent;
+        
+        public EagerOuterProducer(EagerOuterSubscriber<?, ?> parent) {
```
remove `public`? (All the constructors in this file for non-public static classes are marked as public).

akarnokd replied:
It doesn't really matter and it helps me visually tell apart the fields and the constructor in general.

davidmoten commented on src/main/java/rx/internal/operators/OperatorEagerConcatMap.java:
```diff
@@ -0,0 +1,299 @@
+                
+                EagerInnerSubscriber<R> inner;
+                
+                boolean d = done;
```
rename to `outerDone`?

davidmoten commented on src/main/java/rx/internal/operators/OperatorEagerConcatMap.java:
```diff
@@ -0,0 +1,299 @@
+                    
+                    
+                    for (;;) {
+                        d = inner.done;
```
There's a bit of reuse of the `d` boolean here for a different meaning (`innerDone`). Could we use a new variable name for this?

akarnokd commented:
Thanks for the review @davidmoten . Usually I use one letter variables because I find it easier to parse the code and instead of long variable names, I use newlines to separate logical blocks. This way, I don't have to type that many letters and I don't have to wait while Eclipse returns with the content assist (which is blocking by the way and may take hundreds of milliseconds, even on my i7 + SSD).

I took the time and renamed variables as you asked for and added some more unit tests that check the code paths. That being said, I don't plan to do such renames in my contributions very often in the future and encourage anybody to post their PRs with their proposed cleanups/renames.

stealthcode commented on src/main/java/rx/Observable.java:
```diff
@@ -4547,6 +4547,398 @@ public final void onNext(T args) {
+     * @return
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> Observable<T> eagerConcatMap(
```
Should we name these `eagerConcat` instead of `eagerConcatMap`? These overloads do not take a `FuncN` so they aren't really mapping.

davidmoten replied:
Ditto that @stealthcode and for discoverability call it concatEager?

akarnokd commented:
Thanks for the feedback. I've updated the method names to `concatMapEager` and `concatEager`, added the `@Experimental` tags, some javadoc and a missing test for a 2-parameter `concatEager` overload.

akarnokd commented:
Updated with a capacity fix for the `SpscLinkedArrayQueue`.

stevegury commented:
I reviewed in details in the code and I believe it is correct.

But I wonder if subscribing to the sources in a "unbounded mode" is the right thing to do.
At the expense of complexifying the code, I believe it could be possible to split the `request(n)` between the sources. Maybe requesting `1` to all the `m` sources expect the first one, which we request `n - m + 1` ... and so on.

The devil is in the details, and I believe this would significantly complexify the code, but I would like to know what you are thinking about this?

akarnokd commented:
@stevegury This operator consumes source observables in order and doesn't make sense to split any request from downstream. If the downstream requests n and the first is requested n / count, that would hang the sequence because only the first is allowed to emit and it won't emit enough to trigger a new request.

It is possible to use a bounded buffer per source so while they are not consumed, they don't grow indefinitely. However, it means that each of them can produce only `RxRingBuffer.SIZE` elements before stopping and thus the operator wouldn't be eager anymore.

stevegury commented:
@akarnokd sorry I wasn't very clear but what I proposed was roughly what you described (replacing `RxRingBuffer.SIZE` by `1`).

My main concern here is the use of unbounded buffer, which, I think, could be avoided.

akarnokd commented:
@davidmoten What do you think about the bounded buffering and the fact that such sources would act more like a delayed source?

davidmoten commented:
I'd be happy to see bounded buffering (`RxRingBuffer.SIZE` sounds a better default size than 1 to me and would be consistent with the request eagerness of other operators like `merge` for instance). Could we add an overload to allow the specification of buffer size as well?

If this is significant rework I'm content to see this unbounded version documented with its buffering characteristics merged and we can defer the bounding work to another PR.

akarnokd commented:
Done.

davidmoten commented:
Great, thanks @akarnokd ! Lightning quick as always.

stevegury commented:
LGTM :+1:

abersnaze commented:
The three of you seem happy with it so I'm merging it.
